File: codeforces/Contests/1900_Round911_Div2/C.py
cases = int(input())

while cases > 0:
  cases -= 1
  vert = int(input())
  dirs = input()

  parent = {}
  tree = []
  leafs = []
  i = 1
  while vert > 0:
    vert -= 1
    temp = input().split(" ")
    l,r = int(temp[0]), int(temp[1])
    if l == r == 0: leafs.append(i)
    if l != 0: parent[l] = i
    if r != 0: parent[r] = i
    tree.append((l,r))
    i+=1
  parent[1] = -1

  res = 10**9
  for l in leafs:
    modifications = 0

    cur = l
    while True:
      Parent = parent[cur]
      if Parent == -1: break
      ParDirection = dirs[Parent-1]
      if ParDirection == 'U':
        modifications += 1
      else:
        Target = tree[Parent-1][0] if ParDirection == 'L' else tree[Parent-1][1]
        if Target != cur:
          modifications += 1
      cur = Parent

    res = min(res, modifications)
  print(res)

File: codeforces/Contests/1900_Round911_Div2/C.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;

void print_vector(const vector<int>& arr) {
  for (int n : arr)
    cout << n << " ";
  cout << "\n";
}

void solve() {
  int len;
  string dirs;
  cin >> len >> dirs;
  vector<pair<int,int>> tree;
  vi leafs;
  vi parent(len);
  parent[0] = -1;

  for (int i = 0; len-- > 0; i++) {
    int l,r;
    cin >> l >> r;
    l--; r--;
    tree.push_back(make_pair(l,r));
    if (l == -1 && r == -1)
      leafs.push_back(i);
    if (l != -1)
      parent[l] = i;
    if (r != -1)
      parent[r] = i;
  }

  // print_vector(leafs);
  // print_vector(parent);
  // cout << "\n\n";

  int res = INT_MAX;
  for (int l : leafs) {
    int mods = 0;

    int cur = l;
    while (true) {
      int par = parent[cur];
      if (par == -1)
        break;
      char parDir = dirs[par];
      // cout << "cur" << cur << " par" << par << " dir" << parDir << "\n";
      if (parDir == 'U') {
        // cout << "inc\n";
        mods++;
      }
      else {
        int target = (parDir == 'L') ? tree[par].first : tree[par].second;
        if (target != cur) {
          // cout << "inc\n";
          mods++;
        }
      }

      cur = par;
    }

    res = min(res, mods);
  }

  cout << res << "\n";
}

int main() {
  int cases;
  cin >> cases;
  while (cases-- > 0) solve();
}
